lib/common/asar.js

```diff
@@ -39,125 +35,93 @@
-      },
-      isDirectory: function () {
-        return stats.isDirectory
-      },
-      isSymbolicLink: function () {
-        return stats.isLink
-      },
-      isBlockDevice: function () {
-        return false
-      },
-      isCharacterDevice: function () {
-        return false
-      },
-      isFIFO: function () {
-        return false
-      },
-      isSocket: function () {
-        return false
-      }
-    }
+    const { Stats, constants } = require('fs')
+
+    let mode = constants.S_IROTH ^ constants.S_IRGRP ^ constants.S_IRUSR ^ constants.S_IWUSR
+
+    if (stats.isFile) {
+      mode ^= constants.S_IFREG
+    } else if (stats.isDirectory) {
+      mode ^= constants.S_IFDIR
+    } else if (stats.isLink) {
+      mode ^= constants.S_IFLNK
+    }
+
+    const nodeStats = new Stats(
+      1, // dev
+      33188, // mode
+      1, // nlink
+      uid,
+      gid,
+      0, // rdev
+      undefined, // blksize
+      ++nextInode, // ino
+      stats.size,
+      undefined, // blocks,
+      msec(stats.atime), // atim_msec
+      msec(stats.mtime), // mtim_msec
+      msec(stats.ctime), // ctim_msec
+      msec(stats.birthtime) // birthtim_msec
+    )
+
+    nodeStats.isFile = () => stats.isFile
+    nodeStats.isDirectory = () => stats.isDirectory
+    nodeStats.isSymbolicLink = () => stats.isLink
+
+    const forceFalse = (method) => { stats[method] = () => false }
+
+    forceFalse('isBlockDevice')
+    forceFalse('isCharacterDevice')
+    forceFalse('isFIFO')
+    forceFalse('isSocket')
+
+    return nodeStats
   }
 
-  // Create a ENOENT error.
-  const notFoundError = function (asarPath, filePath, callback) {
-    const error = new Error(`ENOENT, ${filePath} not found in ${asarPath}`)
-    error.code = 'ENOENT'
-    error.errno = -2
-    if (typeof callback !== 'function') {
-      throw error
-    }
-    process.nextTick(function () {
-      callback(error)
-    })
-  }
-
-  // Create a ENOTDIR error.
-  const notDirError = function (callback) {
-    const error = new Error('ENOTDIR, not a directory')
-    error.code = 'ENOTDIR'
-    error.errno = -20
-    if (typeof callback !== 'function') {
-      throw error
-    }
-    process.nextTick(function () {
-      callback(error)
-    })
-  }
-
-  // Create a EACCES error.
-  const accessError = function (asarPath, filePath, callback) {
-    const error = new Error(`EACCES: permission denied, access '${filePath}'`)
-    error.code = 'EACCES'
-    error.errno = -13
-    if (typeof callback !== 'function') {
-      throw error
-    }
-    process.nextTick(function () {
-      callback(error)
-    })
-  }
-
-  // Create invalid archive error.
-  const invalidArchiveError = function (asarPath, callback) {
-    const error = new Error(`Invalid package ${asarPath}`)
-    if (typeof callback !== 'function') {
-      throw error
-    }
-    process.nextTick(function () {
-      callback(error)
-    })
+  const types = ['notFound', 'notDir', 'access', 'invalidArchive']
```

Comment:
Can these all be extracted to constants so we don't have magical strings below.  Just something like

```js
const errors = {
  NOT_FOUND: 'notFound',
  NOT_DIR: 'notDir',
  NO_ACCESS: 'access',
  INVALID_ARCHIVE: 'invalidArchive'
}
```